drivers/bbram/bbram_it8xxx2.c

```diff
@@ -90,13 +111,63 @@ static int bbram_it8xxx2_init(const struct device *dev)
 	return 0;
 }
 
-#define BBRAM_INIT(inst)                                                                           \
+#ifdef CONFIG_BBRAM_IT8XXX2_EMUL
+#define BBRAM_DECL_CONFIG(inst)                                                                    \
+	static uint8_t bbram_it8xxx2_emul_buffer_##inst[DT_INST_REG_SIZE(inst)];                   \
+	static const struct bbram_it8xxx2_config bbram_cfg_##inst = {                              \
+		.base_addr = (uintptr_t)bbram_it8xxx2_emul_buffer_##inst,                          \
+		.size = DT_INST_REG_SIZE(inst),                                                    \
+	}
+#else
+#define BBRAM_DECL_CONFIG(inst)                                                                    \
 	static const struct bbram_it8xxx2_config bbram_cfg_##inst = {                              \
 		.base_addr = DT_INST_REG_ADDR(inst),                                               \
 		.size = DT_INST_REG_SIZE(inst),                                                    \
-	};                                                                                         \
+	}
+#endif
+
+#define BBRAM_INIT(inst)                                                                           \
+	BBRAM_DECL_CONFIG(inst);                                                                   \
 	DEVICE_DT_INST_DEFINE(inst, bbram_it8xxx2_init, NULL, NULL, &bbram_cfg_##inst,             \
 			      PRE_KERNEL_1, CONFIG_BBRAM_INIT_PRIORITY,                            \
 			      &bbram_it8xxx2_driver_api);
 
 DT_INST_FOREACH_STATUS_OKAY(BBRAM_INIT);
+
+#define BBRAM_EMUL_INIT(inst)                                                                      \
+	EMUL_DT_INST_DEFINE(inst, NULL, NULL, &bbram_cfg_##inst, NULL, &it8xxx2_emul_backend_api)
+
+#ifdef CONFIG_BBRAM_IT8XXX2_EMUL
+static int it8xxx2_emul_backend_set_data(const struct emul *target, size_t offset, size_t count,
+					 const uint8_t *buffer)
+{
+	const struct bbram_it8xxx2_config *config = target->cfg;
+
+	if (offset + count > config->size) {
+		return -ERANGE;
+	}
+
+	bytecpy(((uint8_t *)config->base_addr + offset), buffer, count);
+	return 0;
+}
+
+static int it8xxx2_emul_backend_get_data(const struct emul *target, size_t offset, size_t count,
+					 uint8_t *buffer)
+{
+	const struct bbram_it8xxx2_config *config = target->cfg;
+
+	if (offset + count > config->size) {
+		return -ERANGE;
+	}
+
+	bytecpy(buffer, ((uint8_t *)config->base_addr + offset), count);
+	return 0;
+}
+
+static const struct emul_bbram_backend_api it8xxx2_emul_backend_api = {
```

Comment:
Would it make more sense for this to be in a separate emulator .c file? I could ask this about some of the other things in this file, but this seemed like the most obvious.

Reply:
Sure, I was trying to keep the change minimal, but you're 100% right that it makes more sense.